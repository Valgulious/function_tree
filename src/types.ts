export type Operator = '+' | '-' | '*' | '/'
export const operators = ['+', '-', '*', '/']
export const LEFT_PARENTHESIS = '('
export const RIGHT_PARENTHESIS = ')'
