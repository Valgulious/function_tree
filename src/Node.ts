export class Node {
    value: string | number
    left: Node | null
    right: Node | null

    constructor(nodeValue: string | number = ' ', nodeLeft: Node | null = null, nodeRight: Node | null = null) {
        this.value = nodeValue
        this.left = nodeLeft
        this.right = nodeRight
    }
}
