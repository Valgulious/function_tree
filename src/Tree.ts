import {Node} from './Node'
import {LEFT_PARENTHESIS, operators, RIGHT_PARENTHESIS} from './types'
// @ts-ignore
import {treeFromArray, treeToASCII} from '../node_modules/treevis/tree'

const enum PARENT {
    LEFT = 'LEFT',
    RIGHT = 'RIGHT'
}

export class Tree {
    root: Node | null

    constructor(treeRoot: Node | null = null) {
        this.root = treeRoot
    }

    get height() {
        return this.getHeight(this.root)
    }

    print() {
        const array = Tree.toArray(this.root)
        treeToASCII(treeFromArray(array))
    }

    printFullTree() {
        const emptyTree = this.getFullEmptyTree(this.height)
        const fullTree = this.getFullTree(this.root)
        const array = Tree.toArray(fullTree)
        treeToASCII(treeFromArray(array))
    }

    private getHeight(root: Node | null): number {
        if (root === null) {
            return 0
        }
        return 1 + Math.max(this.getHeight(root.left), this.getHeight(root.right))
    }

    private getFullEmptyTree(height: number, root?: Node | undefined) {
        if (root === undefined) {
            root = new Node()
        }
        if (height > 1) {
            root.left = new Node()
            root.right = new Node()
            this.getFullEmptyTree(height - 1, root.left)
            this.getFullEmptyTree(height - 1, root.right)
        }
        return root
    }

    private static toArray(root: Node | null): (string | number | null)[] {
        if (root === null) {
            return []
        }
        return this.whileToArray([root])
    }

    private static whileToArray(stack: (Node | null)[]): (string | number | null)[] {
        if (stack.length === 0) {
            return []
        }

        const first = stack[0]
        if (first) {
            return [first.value, ...this.whileToArray([...stack.slice(1), first.left, first.right])]
        } else {
            return [null, ...this.whileToArray([...stack.slice(1)])]
        }

    }

    private getFullTree(root: Node | null): Node {
        if (root === null) {
            return new Node()
        }
        const value = root.value
        const left = root.left !== null ? this.getFullTree(root.left) : null
        const right = root.right !== null ? this.getFullTree(root.right) : null
        return new Node(value, left, right)
    }
}

const expr = (expression: string): Tree => {
    const tokens = expression.split('').filter(el => el !== ' ')
    return new Tree(gen(tokens))
}

const gen = (tokens: string[]): Node => {
    // const root = new Node()
    if (tokens.length === 1) {
        return new Node(tokens[0])
    }
    const lowPriorityIndex = getLowPriorityToken(tokens)
    // Если нет оператора вне скобок
    if (lowPriorityIndex === -1) {
        if (tokens[0] !== LEFT_PARENTHESIS || tokens[tokens.length - 1] !== RIGHT_PARENTHESIS) {
            throw new Error(`Don't know what to do with ${tokens}`)
        }
        return gen(tokens.slice(1, -1))
    }
    const value = tokens[lowPriorityIndex]
    const left = gen(tokens.slice(0, lowPriorityIndex))
    const right = gen(tokens.slice(lowPriorityIndex + 1))
    return new Node(value, left, right)
}

// Возвращает приоритет оператора: чем больше возращаемое значение - тем больше приоритет
const getPriority = (operator: string): number => {
    // Сгруппированы в порядке возрастания приоритета
    const operatorsPriority = [['+', '-'], ['*', '/']]
    return forPriority(operator, operatorsPriority, 0)
}

const forPriority = (operator: string, operatorsPriority: string[][], i: number): number => {
    if (operatorsPriority.length === 0) {
        return -1
    }

    if (operatorsPriority[0].includes(operator)) {
        return i
    }

    return forPriority(operator, operatorsPriority.slice(1), ++i)
}

// Возвращает индекс оператора с наименьшим приоритетом
const getLowPriorityToken = (tokensArray: string[]): number => {
    const lowPriority = Number.MAX_VALUE
    const index = -1
    const parenthesesCount = 0
    return forLow(tokensArray, lowPriority, parenthesesCount, 0, index)
}

const forLow = (tokensArray: string[], lowPriority: number, parenthesesCount: number, i: number, index: number): number => {
    if (tokensArray.length === 0) {
        return index
    }

    const token = tokensArray[0]
    if (operators.includes(token) && parenthesesCount === 0) {
        const tokenPriority = getPriority(token)
        if (tokenPriority < lowPriority) {
            // lowPriority = tokenPriority
            const newIndex = i
            return forLow(tokensArray.slice(1), tokenPriority, parenthesesCount, ++i, newIndex)
        } else {
            return forLow(tokensArray.slice(1), lowPriority, parenthesesCount, ++i, index)
        }
    } else if (token === '(') {
        return forLow(tokensArray.slice(1), lowPriority, ++parenthesesCount, ++i, index)
    } else if (token === ')') {
        return forLow(tokensArray.slice(1), lowPriority, --parenthesesCount, ++i, index)
    } else {
        return forLow(tokensArray.slice(1), lowPriority, parenthesesCount, ++i, index)
    }
}

const treeSolve = (tree: Tree | null): number => {
    if (tree === null) {
        return -1
    }

    return solve(tree.root)
}

const solve = (root: Node | null = null): any => {
    if (root === null) {
        throw new Error('Tree is empty')
    }
    const {value, left, right} = root
    if (!isNaN(parseFloat(value as string))) {
        return parseFloat(value as string)
    } else if (operators.includes(value as string)) {
        if (left === null || right === null) {
            throw new Error(`Need arguments for ${value}`)
        }
        switch (value) {
            case '+':
                return solve(left) + solve(right)
            case '-':
                return solve(left) - solve(right)
            case '*':
                return solve(left) * solve(right)
            case '/':
                return solve(left) / solve(right)
        }
    } else {
        if (typeof value === 'string' && value.trim() === '') {
            throw new Error('Empty node')
        } else {
            throw new Error(`Invalid character ${value}`)
        }
    }
}

const insert = (tree: Tree | null, value: string | number, level: number, position: number): Tree => {
    if (tree === null) {
        return new Tree()
    }

    if (position > Math.pow(2, level - 1)) {
        return tree
    }

    if (tree.root === null) {
        if (level === 1 && position === 1) {
            // this.root = new Node(value)
            return new Tree(new Node(value))
        } else {
            return tree
        }
    }
    return new Tree(whileInsert(tree.root, position, 1, level, value))
}

const whileInsert = (currentNode: Node | null, localPosition: number, currentLevel: number, level: number,
                     value: string | number): Node | null => {

    if (currentNode === null) {
        return null
    }

    if (level === currentLevel) {
        return new Node(value, currentNode.left, currentNode.right)
    }

    const middle = Math.pow(2, level - currentLevel - 1)
    // Если нужна позиция больше чем середина (для текущего уровня)
    if (localPosition > middle) {
        if (currentNode.right === null) {
            // Если справа пустая Node, но она нам и нужна
            if (level === currentLevel + 1 && localPosition <= 2) {
                return new Node(currentNode.value, currentNode.left, new Node(value))
            }
            return null
        }
        const newLocalPosition = localPosition - middle
        const right = whileInsert(currentNode.right, newLocalPosition, ++currentLevel, level, value)
        return new Node(currentNode.value, currentNode.left, right)
    } else {
        if (currentNode.left === null) {
            if (level === currentLevel + 1) {
                return new Node(currentNode.value, new Node(value), currentNode.right)
            }
            return null
        }
        const left = whileInsert(currentNode.left, localPosition, ++currentLevel, level, value)
        return new Node(currentNode.value, left, currentNode.right)
    }
}

const remove = (tree: Tree, level: number, position: number): Tree => {
    if (tree.root === null) {
        return tree
    }

    return new Tree(whileRemove(tree.root, tree.root, position, 1, null, level))
}

const whileRemove = (currentNode: Node | null, parent: Node | null, localPosition: number,
                     currentLevel: number, direction: PARENT | null, level: number): Node | null => {

    if (currentNode === null) {
        console.log('curr = null')
        return null
    }

    if (level === currentLevel) {
        return null
    }

    const middle = Math.pow(2, level - currentLevel - 1)
    if (localPosition > middle) {
        if (currentNode.right === null) {
            return currentNode
        }
        const newLocalPosition = localPosition - middle
        const right = whileRemove(currentNode.right, currentNode, newLocalPosition, ++currentLevel, PARENT.RIGHT, level)
        return new Node(currentNode.value, currentNode.left, right)
    } else {
        if (currentNode.left === null) {
            return currentNode
        }
        const left = whileRemove(currentNode.left, currentNode, localPosition, ++currentLevel, PARENT.LEFT, level)
        return new Node(currentNode.value, left, currentNode.right)
    }
}

export {expr, treeSolve, insert, remove}