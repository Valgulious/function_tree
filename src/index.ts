import {Tree} from './Tree'

const arg = process.argv.slice(2)

if (arg.length === 1) {
    const tree = new Tree().expr(arg[0])
    // tree.expr(arg[0])
    tree.print()
    console.log(tree.solve())
    process.exit(0)
}

const expr = '1 + 1 * 2'
const tree = new Tree().expr(expr)
// tree.expr(expr)
tree.print()
tree.remove(2, 1)
tree.insert(2, 2, 1)
console.log(tree.solve())
