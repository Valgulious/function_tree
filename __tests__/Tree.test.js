import {Tree} from "../src/Tree"

describe('Mathematical expressions', () => {
	test.each([
		['1 + 1', 2],
		['2 + 2 * 2', 6],
		['(1 + 2) * ( 3 - 4 ) * 9', -27],
		['((1 + 1) / 2 - (5 * (3 + 1)))', -19],
		['1 / 0', Infinity],
	])('%s === %d', (expr, result) => {
		const tree = new Tree().expr(expr)
		// tree.expr(expr)
		expect(tree.solve()).toBe(result)
	})
})

describe('Inserting nodes', () => {
	let tree

	beforeEach(() => {
		tree = new Tree()
	})

	afterEach(() => {
		expect(tree).toMatchSnapshot()
	})

	it('should be simple and return 2', () => {
		tree.insert('+', 1, 1)
		tree.insert('1', 2, 1)
		tree.insert(1, 2, 2)
		expect(tree.solve()).toBe(2)
	})

	it('should throw error on empty tree', () => {
		expect(() => tree.solve()).toThrowError(/empty/i)
	})

	it('should parse expression, insert some nodes, and return 88', () => {
		tree.expr('1 + 2 * (3 + 4)')
		// replace 1 on 11
		tree.insert(11, 2, 1)
		// replace 2 on +, add arguments 5 and 6
		tree.insert('+', 3, 3)
		tree.insert(5, 4, 5)
		tree.insert(6, 4, 6)
		expect(tree.solve()).toBe(88)
	})

	it('should replace root node operator', () => {
		tree.insert('+', 1, 1)
		tree.insert('1', 2, 1)
		tree.insert(1, 2, 2)
		tree.insert('-', 1, 1)
		expect(tree.solve()).toBe(0)
	})

	it('should not insert value into level 2 on empty tree', () => {
		const result = tree.insert('1', 2, 1)
		expect(result).toBe(-1)
	})
})


describe('Deleting nodes', () => {
	let tree

	beforeEach(() => {
		tree = new Tree()
	})

	afterEach(() => {
		expect(tree).toMatchSnapshot()
	})

	it('should remove one node', () => {
		tree.expr('1 + 1')
		tree.remove(2, 1)
	})

	it('should remove leaf', () => {
		tree.expr('1 + 2 * ( 3 + 4 )')
		tree.remove(4, 8)
	})

	it('should remove node and its children', () => {
		tree.expr('1 + 2 * ( 3 + 4 )')
		tree.remove(2, 2)
	})

	it('should remove root of tree', () => {
		tree.expr('1 + 2 * ( 3 + 4 )')
		tree.remove(1, 1)
	})
})
