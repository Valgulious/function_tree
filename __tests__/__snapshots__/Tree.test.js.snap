// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deleting nodes should remove leaf 1`] = `
Tree {
  "getLowPriorityToken": [Function],
  "getPriority": [Function],
  "root": Node {
    "left": Node {
      "left": null,
      "right": null,
      "value": "1",
    },
    "right": Node {
      "left": Node {
        "left": null,
        "right": null,
        "value": "2",
      },
      "right": Node {
        "left": Node {
          "left": null,
          "right": null,
          "value": "3",
        },
        "right": null,
        "value": "+",
      },
      "value": "*",
    },
    "value": "+",
  },
}
`;

exports[`Deleting nodes should remove node and its children 1`] = `
Tree {
  "getLowPriorityToken": [Function],
  "getPriority": [Function],
  "root": Node {
    "left": Node {
      "left": null,
      "right": null,
      "value": "1",
    },
    "right": null,
    "value": "+",
  },
}
`;

exports[`Deleting nodes should remove one node 1`] = `
Tree {
  "getLowPriorityToken": [Function],
  "getPriority": [Function],
  "root": Node {
    "left": null,
    "right": Node {
      "left": null,
      "right": null,
      "value": "1",
    },
    "value": "+",
  },
}
`;

exports[`Deleting nodes should remove root of tree 1`] = `
Tree {
  "getLowPriorityToken": [Function],
  "getPriority": [Function],
  "root": null,
}
`;

exports[`Inserting nodes should be simple and return 2 1`] = `
Tree {
  "getLowPriorityToken": [Function],
  "getPriority": [Function],
  "root": Node {
    "left": Node {
      "left": null,
      "right": null,
      "value": "1",
    },
    "right": Node {
      "left": null,
      "right": null,
      "value": 1,
    },
    "value": "+",
  },
}
`;

exports[`Inserting nodes should not insert value into level 2 on empty tree 1`] = `
Tree {
  "getLowPriorityToken": [Function],
  "getPriority": [Function],
  "root": null,
}
`;

exports[`Inserting nodes should parse expression, insert some nodes, and return 88 1`] = `
Tree {
  "getLowPriorityToken": [Function],
  "getPriority": [Function],
  "root": Node {
    "left": Node {
      "left": null,
      "right": null,
      "value": 11,
    },
    "right": Node {
      "left": Node {
        "left": Node {
          "left": null,
          "right": null,
          "value": 5,
        },
        "right": Node {
          "left": null,
          "right": null,
          "value": 6,
        },
        "value": "+",
      },
      "right": Node {
        "left": Node {
          "left": null,
          "right": null,
          "value": "3",
        },
        "right": Node {
          "left": null,
          "right": null,
          "value": "4",
        },
        "value": "+",
      },
      "value": "*",
    },
    "value": "+",
  },
}
`;

exports[`Inserting nodes should replace root node operator 1`] = `
Tree {
  "getLowPriorityToken": [Function],
  "getPriority": [Function],
  "root": Node {
    "left": Node {
      "left": null,
      "right": null,
      "value": "1",
    },
    "right": Node {
      "left": null,
      "right": null,
      "value": 1,
    },
    "value": "-",
  },
}
`;

exports[`Inserting nodes should throw error on empty tree 1`] = `
Tree {
  "getLowPriorityToken": [Function],
  "getPriority": [Function],
  "root": null,
}
`;
